src/mpyl/projects/versioning.py

```diff
@@ -0,0 +1,149 @@
+implements the `upgrade` method. This class should then be added to the `UPGRADERS`
+list in this module.
+
+"""
+import copy
+import pkgutil
+from abc import ABC
+from pathlib import Path
+from typing import Optional, Generator
+
+from deepdiff import DeepDiff
+from ruamel.yaml.compat import ordereddict
+
+from ..utilities.yaml import yaml_to_string, load_for_roundtrip
+
+VERSION_FIELD = "mpylVersion"
+
+
+def get_releases() -> list[str]:
+    embedded_releases = pkgutil.get_data(__name__, "releases/releases.txt")
+    if not embedded_releases:
+        raise ValueError("File releases/releases.txt not found in package")
+    releases = embedded_releases.decode("utf-8").strip().splitlines()
+    return list(reversed(releases))
+
+
+def get_latest_release() -> str:
+    return get_releases()[0]
+
+
+class Upgrader(ABC):
+    """Base class for upgrade scripts"""
+
+    target_version: str
+
+    def upgrade(self, previous_dict: ordereddict) -> ordereddict:
+        return previous_dict
+
+
+class Upgrader11(Upgrader):
+    target_version = "1.0.11"
+
+    def upgrade(self, previous_dict: ordereddict) -> ordereddict:
+        return previous_dict
+
+
+class Upgrader10(Upgrader):
+    target_version = "1.0.10"
+
+    def upgrade(self, previous_dict: ordereddict) -> ordereddict:
+        if found_deployment := previous_dict.get("deployment", {}):
+            existing_namespace = found_deployment.get("namespace", None)
+            if not existing_namespace:
+                previous_dict["deployment"].insert(
+                    0, "namespace", previous_dict["name"]
+                )
+
+        return previous_dict
+
+
+class Upgrader9(Upgrader):
+    target_version = "1.0.9"
+
+    def upgrade(self, previous_dict: ordereddict) -> ordereddict:
+        previous_dict.insert(3, VERSION_FIELD, self.target_version)
+        return previous_dict
+
+
+class Upgrader8(Upgrader):
+    target_version = "1.0.8"
+
+    def upgrade(self, previous_dict: ordereddict) -> ordereddict:
+        return previous_dict
+
+
+UPGRADERS = [Upgrader8(), Upgrader9(), Upgrader10(), Upgrader11()]
+
+
+def get_entry_upgrader_index(
+    current_version: str, upgraders: list[Upgrader]
+) -> Optional[int]:
+    next_index = next(
+        (i for i, v in enumerate(upgraders) if v.target_version == current_version),
+        None,
+    )
+    return next_index
+
+
+def __get_version(project: dict) -> str:
+    return project.get(VERSION_FIELD, "1.0.8")
+
+
+def upgrade_to_latest(to_upgrade: ordereddict, upgraders: list[Upgrader]):
+    versie = __get_version(to_upgrade)
+    upgrade_index = get_entry_upgrader_index(versie, upgraders)
+    if upgrade_index is None:
+        return to_upgrade
+
+    upgraded = to_upgrade
+    for i in range(upgrade_index, len(upgraders)):
+        upgrader = upgraders[i]
+        before_upgrade = copy.deepcopy(upgraded)
+        upgraded = upgrader.upgrade(copy.deepcopy(before_upgrade))
```

Comment:
I think you've refactored this and forgot to remove the old `copy.deepcopy` because now it's done twice on the same object.
Also, would the copying be needed if you just start with an clean empty dict for `upgraded` on line 103?

Reply:
We do need the double copy. Otherwise `before_upgrade` will point to the some object as  `upgraded` when we do the diff.

Not sure how we can start with an empty dict. If there's nothing to upgrade we should do an identity transform, not return an empty dictionary.

Reply:
I thought that was a js thing 😅